File: frontend/src/services/aviator.ts
import api from './api';

export interface AviatorRoundHistoryItem {
  id: string;
  nonce: number;
  server_seed_hash: string;
  server_seed?: string | null;
  crash_multiplier?: number | null;
  status: string;
  betting_started_at: string;
  flight_started_at?: string | null;
  crashed_at?: string | null;
  settled_at?: string | null;
}

export interface AviatorBetHistoryItem {
  id: string;
  user_id: string;
  round_id: string;
  slot: number;
  amount: number;
  auto_cashout?: number | null;
  status: string;
  cashout_multiplier?: number | null;
  payout: number;
  cashed_out_at?: string | null;
  created_at: string;
}

export interface AviatorFairnessData {
  round_id: string;
  nonce: number;
  server_seed_hash: string;
  server_seed?: string | null;
  crash_multiplier?: number | null;
  status: string;
  verification_note: string;
}

export const aviatorService = {
  async getHistory(limit = 20): Promise<AviatorRoundHistoryItem[]> {
    const res = await api.get('/aviator/history', { params: { limit } });
    return res.data.data;
  },

  async getMyBets(limit = 20): Promise<AviatorBetHistoryItem[]> {
    const res = await api.get('/aviator/my-bets', { params: { limit } });
    return res.data.data;
  },

  async getFairness(roundId: string): Promise<AviatorFairnessData> {
    const res = await api.get(`/aviator/fairness/${roundId}`);
    return res.data.data;
  },
};

File: backend/app/models/aviator.py
"""
Aviator database models — persisted round and bet history.
"""

import uuid
import enum
from datetime import datetime, timezone
from sqlalchemy import (
    Column, BigInteger, Integer, String, Float, DateTime,
    Enum as SAEnum, ForeignKey, Boolean,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from ..database import Base


class AviatorRoundStatus(str, enum.Enum):
    BETTING = "BETTING"
    FLYING = "FLYING"
    CRASHED = "CRASHED"
    SETTLED = "SETTLED"


class AviatorBetStatus(str, enum.Enum):
    ACTIVE = "ACTIVE"
    CASHED_OUT = "CASHED_OUT"
    LOST = "LOST"


class AviatorRound(Base):
    __tablename__ = "aviator_rounds"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    nonce = Column(Integer, nullable=False)
    server_seed_hash = Column(String(64), nullable=False)
    server_seed = Column(String(64), nullable=True)          # revealed after crash
    crash_multiplier = Column(Float, nullable=True)           # revealed after crash
    status = Column(
        SAEnum(AviatorRoundStatus, name="aviator_round_status"),
        nullable=False,
        default=AviatorRoundStatus.BETTING,
    )
    betting_started_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(timezone.utc))
    flight_started_at = Column(DateTime(timezone=True), nullable=True)
    crashed_at = Column(DateTime(timezone=True), nullable=True)
    settled_at = Column(DateTime(timezone=True), nullable=True)

    bets = relationship("AviatorBet", back_populates="round", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<AviatorRound id={self.id} nonce={self.nonce} crash={self.crash_multiplier} status={self.status}>"


class AviatorBet(Base):
    __tablename__ = "aviator_bets"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    round_id = Column(UUID(as_uuid=True), ForeignKey("aviator_rounds.id", ondelete="CASCADE"), nullable=False, index=True)
    slot = Column(Integer, nullable=False)                    # 1 or 2
    amount = Column(BigInteger, nullable=False)               # paise
    auto_cashout = Column(Float, nullable=True)               # target multiplier
    status = Column(
        SAEnum(AviatorBetStatus, name="aviator_bet_status"),
        nullable=False,
        default=AviatorBetStatus.ACTIVE,
    )
    cashout_multiplier = Column(Float, nullable=True)
    payout = Column(BigInteger, nullable=True, default=0)
    cashed_out_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(timezone.utc))

    round = relationship("AviatorRound", back_populates="bets")

    def __repr__(self):
        return f"<AviatorBet id={self.id} user={self.user_id} slot={self.slot} status={self.status}>"
